### DISCUSSION TITLE
Restarting Language Servers

### DISCUSSION THREAD

**Discussion by Paracelsus-Rose** [VonHeikemen/lsp-zero.nvim#63] [Q&A] [answered]
I use a fair amount of file generators while working.  Usually this causes the need to restart language servers while developing. Namely, Restarting Volar, and restarting EsLint. 

In VsCode I would ctrl shift P and select Restart Volar or Restart EsLint.  This causes the language servers to reread the root folder and build suggestions for the new directories and files created.

Currently, When I issue LspRestart while using LspZero, It seems to fail in some form.  My current save keybind is ```<cmd> w! | LspZeroFormat <CR>``` which works fine prior to restarting the servers.  Once I do it seems there is an issue reattaching to the current buffer.

Not sure if this issue is outside the scope of LspZero, and if it is I apologize for starting the discussion here.  

Currently I work around this by running a few terminals in splits, using my generator in one terminal, then just exiting nvim completely and reopening it to re initialize the LspServers once files are generated.  

Let me know if you are aware of any wizardry like it seems you always are, or a way to remedy this.  To be honest, with harpoon, and telescope, I really do not "mind" reopening nvim each time.  Just for some of the people I share my config with, I know it is a bit of a deterrent for them. 

Once again thank you for all of the work you do, it is such a benefit to have people like you in the community. Cheers.

**Comment by VonHeikemen** [comment]
I think is an issue with lsp-zero, `lspconfig` does not like when someone delays the setup of the lsp servers.

After `LspRestart` I usually just execute the command `:edit` if the language server fails to attach to a buffer. This will trigger the autocommand neovim uses to attach the lsp.

Edit:

Can you test if this happens with the preset `manual-setup`?

```lua
local lsp = require('lsp-zero')
lsp.preset('manual-setup')

lsp.setup_servers({
  'tsserver',
  'volar',
  -- etc....
})

lsp.setup()
```

**Reply by Paracelsus-Rose** [reply]
```lua
local lsp = require("lsp-zero")

lsp.preset("manual-setup")

local prefs = require("plugins.lsp.settings.preferences")
lsp.set_preferences(prefs.preferences)

local ensure_installed = require("plugins.lsp.servers.server_list").lsp_ensure_installed
lsp.ensure_installed(ensure_installed)

local sumneko_opts = require("plugins.lsp.servers.sumneko_lua")
local rust_ops = require("plugins.lsp.servers.rust")
local ts_opts = require("plugins.lsp.servers.tsserver")
local eslint_opts = require("plugins.lsp.servers.eslint")
local css_opts = require("plugins.lsp.servers.css")
local json_opts = require("plugins.lsp.servers.json")

lsp.configure("jsonls", {
    settings = json_opts.settings,
})

lsp.configure("sumneko_lua", {
    settings = sumneko_opts.settings,
})

lsp.configure("eslint", {
    on_attach = function(client, bufnr)
        client.server_capabilities.documentFormattingProvider = true
        local function buf_set_option(...)
            vim.api.nvim_buf_set_option(bufnr, ...)
        end

        buf_set_option("omnifunc", "v:lua.vim.lsp.omnifunc")
    end,
    settings = eslint_opts.settings,
})

lsp.configure("cssls", {
    on_attach = function(client, bufnr)
        client.server_capabilities.documentFormattingProvider = true
        client.server_capabilities.documentRangeFormattingProvider = true
    end,

    settings = css_opts.settings,
})

local rust_lsp = lsp.build_options("rust_analyzer", rust_ops.rust_server_settings)

local typescript_lsp = lsp.build_options("tsserver", ts_opts.ts_settings)

lsp.setup()

require("rust-tools").setup({
    tools = rust_ops.tools,
    server = rust_lsp,
})

require("typescript").setup({
    debug = false,
    disable_commands = false,
    server = typescript_lsp,
})

local cmp = require("cmp")

vim.opt.completeopt = { "menu", "menuone", "noselect" }

local cmp_conf = require("plugins.lsp.settings.cmp").cmp

cmp.setup(cmp_conf)

cmp.setup.cmdline("/", {
    preselect = "none",
    sources = {
        { name = "treesitter" },
        { name = "buffer", keyword_length = 2 },
        { name = "nvim_lsp", keyword_length = 3 },
    },
})

cmp.setup.cmdline(":", {
    preselect = "none",
    sources = cmp.config.sources({
        { name = "path", keyword_length = 2 },
    }, {
        { name = "cmdline", keyword_length = 2 },
    }),
})

vim.diagnostic.config(prefs.vim_lsp_pref)

```

This is my current lsp-zero config trying to set it up with manual_setup.  Does this look correct for what you want tested?  Not sure if since I am using build_options if I need to run setup_servers also.  Thanks for the prompt response.

**Comment by VonHeikemen** [comment]
`.setup_servers()` is just a utility function that uses `.configure()` under the hood. Is just so users don't feel the need to create `for loops`.

The config looks good. Did it fix the issue with the LSP server restart?

**Reply by Paracelsus-Rose** [reply]
Apologize for the late reply, got swamped with work this week.

This fixed it. I just was able to generate a new file externally, restart the servers and find the file for import.  Thanks so much for the help.  This seems to have solved the issue though. I will test it all night tonight and if I find any issues report back here.

Side discussion. I can start a new one or continue it here. 

![image](https://user-images.githubusercontent.com/103412754/204689997-0b71c77b-563c-4958-80d9-d7f44fa699d9.png)

Currently I have ``` ['<C-s>'] = { '<cmd> w! | LspZeroFormat <CR>' },``` set in my keymaps, and It has always worked. Currently, when I trigger a save, I get the shown popup message, but it still seems to be formatting correctly with no issues.  Not sure if this is Lsp-Zero Related or related to Noice.  Once again, you are are beyond appreciated for all you do.

**Reply by Paracelsus-Rose** [reply]
After further investigation, it would seem that it is only EsLint that is not reattaching itself, on either configuration, either lsp-compe, or manual how you have shown.

Not sure how to check into it from here.  Entering into realms I do not know.  Will do my best to look around and reply back.

These are the EsLint settings I am passing to lsp.configure() If it helps.
```lua
local M = {}

M.settings = {
    codeAction = {
        disableRuleComment = {
            enable = true,
            location = 'separateLine',
        },
        showDocumentation = {
            enable = true,
        },
    },
    codeActionOnSave = {
        enable = true,
        mode = 'all',
    },
    format = true,
    nodePath = '',
    onIgnoredFiles = 'off',
    packageManager = 'yarn',
    quiet = false,
    rulesCustomizations = {},
    run = 'onType',
    useESLintClass = false,
    validate = 'on',
    workspaceFolder = {
        mode = 'location',
    },
}

return M
```

**Reply by VonHeikemen** [reply]
You could also try using `lspconfig` directly. Use `.build_options` to get the config and pass that to lspconfig.

```lua
local eslint_lsp = lsp.build_options('eslint', {
  settings = eslint_opts.settings,
  on_attach = function(client, bufnr)
    client.server_capabilities.documentFormattingProvider = true
    local function buf_set_option(...)
      vim.api.nvim_buf_set_option(bufnr, ...)
    end

    buf_set_option("omnifunc", "v:lua.vim.lsp.omnifunc")
  end,
})

require('lspconfig').eslint.setup(eslint_lsp)
```

**Reply by Paracelsus-Rose** [reply]
![image](https://user-images.githubusercontent.com/103412754/205223314-4b439127-c6fa-4888-879a-3381eee1cdb3.png)

Still unable to attach after LSP Restart, using this config below.  Appreciate the help thus far, and I do apologize if this is not so much an lsp-zero related issue or not, at this point I am unsure.

```lua 
local lsp = require("lsp-zero")

lsp.preset("manual-setup")

local prefs = require("plugins.lsp.settings.preferences")
lsp.set_preferences(prefs.preferences)

local ensure_installed = require("plugins.lsp.servers.server_list").lsp_ensure_installed
lsp.ensure_installed(ensure_installed)

local css_opts = require("plugins.lsp.servers.css")
local eslint_opts = require("plugins.lsp.servers.eslint")
local json_opts = require("plugins.lsp.servers.json")
local sumneko_opts = require("plugins.lsp.servers.sumneko_lua")
local rust_ops = require("plugins.lsp.servers.rust")
local ts_opts = require("plugins.lsp.servers.tsserver")

lsp.configure("jsonls", {
    settings = json_opts.settings,
})

lsp.configure("sumneko_lua", {
    settings = sumneko_opts.settings,
})

local eslint_lsp = lsp.build_options('eslint', {
    settings = eslint_opts.settings,
    on_attach = function(client, bufnr)
        client.server_capabilities.documentFormattingProvider = true
        local function buf_set_option(...)
            vim.api.nvim_buf_set_option(bufnr, ...)
        end

        buf_set_option("omnifunc", "v:lua.vim.lsp.omnifunc")
    end,
})

require('lspconfig').eslint.setup(eslint_lsp)


lsp.configure("cssls", {
    on_attach = function(client, bufnr)
        client.server_capabilities.documentFormattingProvider = true
        client.server_capabilities.documentRangeFormattingProvider = true
    end,

    settings = css_opts.settings,
})

local rust_lsp = lsp.build_options("rust_analyzer", rust_ops.rust_server_settings)

local typescript_lsp = lsp.build_options("tsserver", ts_opts.ts_settings)

lsp.setup_servers({
    'bashls',
    'html',
    'pyright',
    'taplo',
    'vimls',
    'volar',
    'zk'
})
lsp.setup()

require("rust-tools").setup({
    tools = rust_ops.tools,
    server = rust_lsp,
})

require("typescript").setup({
    debug = false,
    disable_commands = false,
    server = typescript_lsp,
})

local cmp = require("cmp")

vim.opt.completeopt = { "menu", "menuone", "noselect" }

local cmp_conf = require("plugins.lsp.settings.cmp").cmp

cmp.setup(cmp_conf)

cmp.setup.cmdline("/", {
    preselect = "none",
    sources = {
        { name = "treesitter" },
        { name = "buffer", keyword_length = 2 },
        { name = "nvim_lsp", keyword_length = 3 },
    },
})

cmp.setup.cmdline(":", {
    preselect = "none",
    sources = cmp.config.sources({
        { name = "path", keyword_length = 2 },
    }, {
        { name = "cmdline", keyword_length = 2 },
    }),
})

vim.diagnostic.config(prefs.vim_lsp_pref)


```

**Comment by VonHeikemen** [comment, marked as answer]
Looks like the issue is with `eslint` and `lspconfig`. If you can reproduce this bug with a minimal config it might be worth opening an issue in [nvim-lspconfig](https://github.com/neovim/nvim-lspconfig).

You can make a new config and use it with the `-u` flag.

```sh
nvim -u /path/to/test-config.lua
```

You can use this config to test.

```lua
-- exclude default paths/config
vim.opt.runtimepath:remove(vim.fn.expand('~/.config/nvim'))
vim.opt.packpath:remove(vim.fn.expand('~/.local/share/nvim/site'))

-- add new path
local config_dir = '/tmp/nvim-config'

vim.opt.runtimepath:append(config_dir)
vim.opt.packpath:append(config_dir)

-- Install minpac
local install_path = config_dir .. '/pack/minpac/opt/minpac'
local install_plugins = false

if vim.fn.empty(vim.fn.glob(install_path)) > 0 then
  vim.cmd('!git clone https://github.com/k-takata/minpac ' .. install_path)
  install_plugins = true
end

local pack_init = function()
  vim.cmd('packadd minpac')

  vim.call('minpac#init', {dir = config_dir})
  local Plug = vim.fn['minpac#add']

  -- Plugin manager
  Plug('k-takata/minpac', {type = 'opt'})

  -- LSP Support
  Plug('neovim/nvim-lspconfig')
  Plug('williamboman/mason.nvim')
  Plug('williamboman/mason-lspconfig.nvim')
end

-- Create commands to manage plugins
local command = vim.api.nvim_create_user_command
local pack_cmd = function(action)
  return function() pack_init(); vim.call(action) end
end

command('PackUpdate', pack_cmd('minpac#update'), {desc = 'Update plugins'})
command('PackClean', pack_cmd('minpac#clean'), {desc = 'Remove unlisted plugins'})
command('PackStatus', pack_cmd('minpac#status'), {desc = "Open minpac's progress window"})

if install_plugins then
  vim.cmd('PackUpdate')
  return
end

require('mason').setup({})
require('mason-lspconfig').setup({})

require('lspconfig').eslint.setup({
  on_attach = function(client, bufnr)
    print('eslint started')
  end,
})
```

**Reply by Paracelsus-Rose** [reply]
https://github.com/neovim/nvim-lspconfig/issues/2289

**Reply by Paracelsus-Rose** [reply]
Once again you were right. Submitted the bug report and it was fixed. Things are working swimmingly again. Last thing is this message popping up. As for now I disabled error reporting to not see it on save. Any suggestions?

![image](https://user-images.githubusercontent.com/103412754/206327268-bbb697c0-2d6e-493c-8d57-95f03297c1d6.png)

**Reply by VonHeikemen** [reply]
Is this an `eslint` thing? The LSP server for eslint doesn't implement formatting in a "traditional" way. If you want to format with eslint, you need to use the command `EslintFixAll`.

**Reply by Paracelsus-Rose** [reply]
This was happening in any file. Lua, json etc., Did not seem eslint related. As of now though it has seemed to have stopped happening.  Looked through changelogs and do not see anything of relevance. Really appreciate your help in all of this. You got my neovim working better than vscode ever did. 

Will say the only thing I am having an issue trying to replicate, and this could be from just pure misuse, would be to trigger suggestions while inside a type.  Sometimes you will say set a type to 

```ts
type Foo = 'foo' | 'bar'

// Then elsewhere
const foo: Foo = ' '
```

In vscode if you were to put your cursor between the ' ' and hit ctrl space. A popup appears with 'foo' and 'bar' in it.  Do you know if this is possible? I have autocompletion and snippets working fine. Just seems I cant figure out triggering suggest.

![image](https://user-images.githubusercontent.com/103412754/207735377-502e07db-b0c8-419b-89ed-4bcfe4f6e9bf.png)

**Reply by VonHeikemen** [reply]
Use `Ctrl+e` or `Tab`.

`Ctrl+e` just toggles the completion menu.

`Tab` can also trigger the completion, but it needs at least one non-whitespace character before the cursor.

**Reply by Paracelsus-Rose** [reply]
It would seem I had something else not allowing the keybind.  Believe it is another plugin. Remapped what you have in under the hood for ctrl e and tab to other keys, and it is working phenomenally.   You are such a blessing to the community. Cheers boss. Thank you for everything along the way.
